crates/rethnet_eth/src/transaction/signed/eip155.rs

```diff
@@ -0,0 +1,86 @@
+use bytes::Bytes;
+use revm_primitives::{keccak256, Address, B256, U256};
+
+use crate::{
+    signature::{Signature, SignatureError},
+    transaction::{kind::TransactionKind, request::EIP155TransactionRequest},
+};
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+#[cfg_attr(
+    feature = "fastrlp",
+    derive(open_fastrlp::RlpEncodable, open_fastrlp::RlpDecodable)
+)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct EIP155SignedTransaction {
+    pub nonce: u64,
+    pub gas_price: U256,
+    pub gas_limit: u64,
+    pub kind: TransactionKind,
+    pub value: U256,
+    pub input: Bytes,
+    pub signature: Signature,
+}
+
+impl EIP155SignedTransaction {
+    pub fn hash(&self) -> B256 {
+        keccak256(&rlp::encode(self))
+    }
+
+    /// Recovers the Ethereum address which was used to sign the transaction.
+    pub fn recover(&self) -> Result<Address, SignatureError> {
+        self.signature
+            .recover(EIP155TransactionRequest::from(self).hash())
+    }
+
+    pub fn chain_id(&self) -> u64 {
+        (self.signature.v - 35) / 2
```

Comment:
Is this correct if the parity is odd? Would be good to have some tests here

Reply:
I'm happy to add tests, but would you mind if I do that in https://github.com/NomicFoundation/hardhat/pull/4300/files/3c85628c4f99db9b9274efecce0558d937756f5b#diff-6ba4eabcc2f6a81547196c40a9e9a9ecdc9a53d1a3b6a75012953e6f1960fe0b, since that adds a bunch of tests and would cause conflicts if I also add a module here 😅

Reply:
Sure, sounds good!